### PR TITLE
Fix examples client

### PR DESCRIPTION
**Pull request by dhoomakethu**
* Fix examples
* Fix client unable to send to messages over the socket `AttributeError("'NoneType' object has no attribute 'send'")`

### CODE REVIEW

File: examples/server_async.py
Comment (janiversen) [inline examples/server_async.py:289]: This is quite a complicated way of doing that. please use the comm_defaults where the transport defaults are stored.

File: examples/server_sync.py
Comment (janiversen) [inline examples/server_sync.py:183]: Where did you see problems with this code ? I run the server daily and have so far not had any problems.

File: examples/server_sync.py
Comment (janiversen) [inline examples/server_sync.py:288]: Same as before please use comm_defaults.

File: pymodbus/client/base.py
Comment (janiversen) [inline pymodbus/client/base.py:81]: That change causes problems. Where did you see the problem with the code as it is.

ClientDecoder does not take an extra argument.

File: examples/client_sync.py
Comment (janiversen) [reply examples/client_sync.py:125]: Whenever start() is called stop() should also be called, if that causes a problem it needs to be fixed in the library.

**Review by janiversen — CHANGES_REQUESTED** [review]
I thought you wrote something about "Don't want to step in to your refactoring work." !

But seems you have decided not to follow that.

It is nice to see you active again, and a larger number of your changes are good....BUT some of the changes are made for a purpose (like using comm_defaults) and it is less nice to have those changed.

File: pymodbus/client/base.py
Comment (dhoomakethu) [reply pymodbus/client/base.py:81]: I am having issues with the latest dev branch. Can't get anything working and these changes are what that made it working. I am not sure of your setup but I would request you use a new venv and try.

File: examples/client_sync.py
Comment (dhoomakethu) [reply examples/client_sync.py:125]:
```
/Users/sanjay/.virtualenvs/pymodbus/bin/python /Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py --com tcp --framer socket --port 5020
22:29:33 INFO  client_sync:47 ### Create client object
22:29:33 INFO  client_sync_basic_calls:16 ### Reading Coil
22:29:33 INFO  client_sync_basic_calls:22 ### Reading Coils to get bit 5
22:29:33 INFO  client_sync_basic_calls:28 ### Write true to coil bit 0 and read to verify
22:29:33 INFO  client_sync_basic_calls:36 ### Write true to multiple coils 1-8
22:29:33 INFO  client_sync_basic_calls:49 ### Write False to address 1-8 coils
22:29:33 INFO  client_sync_basic_calls:60 ### Reading discrete input, Read address:0-7
22:29:33 INFO  client_sync_basic_calls:69 ### write holding register and read holding registers
22:29:33 INFO  client_sync_basic_calls:77 ### write holding registers and read holding registers
22:29:33 INFO  client_sync_basic_calls:85 ### write read holding registers
22:29:33 INFO  client_sync_basic_calls:103 ### read input registers
Traceback (most recent call last):
  File "/Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py", line 119, in <module>
    run_sync_client(demonstrate_calls)
  File "/Users/sanjay/repo/pymodbus/examples/client_sync.py", line 125, in run_sync_client
    client.stop()
AttributeError: 'ModbusTcpClient' object has no attribute 'stop'
```

File: pymodbus/client/base.py
Comment (janiversen) [reply pymodbus/client/base.py:81]: A new venv have nothing to do with this!!! you can see that all tests run ok, as well as pylint etc. so do not tell me that nothing is working. Please remember that CI makes a new venv every time.

Apart from that, you should look deeper into your changes. You e.g. removed a .stop in the example, but the real problem is in the library (solved on dev_client).

For sure there are momentary problems, because I focus on the changes, while securing that I have a GREEN CI before merging anything, but the examples are a bit later, when the client is stabilised.

I suggest you use an older dev.

File: examples/server_sync.py
Comment (dhoomakethu) [reply examples/server_sync.py:183]:
```
/Users/sanjay/.virtualenvs/pymodbus/bin/python /Users/sanjay/repo/pymodbus/examples/server_sync.py --log debug --comm serial --framer rtu --port /tmp/ptyp0
22:32:23 INFO  server_sync:67 ### Create datastore
22:32:23 INFO  server_sync:142 ### start server
Traceback (most recent call last):
  File "/Users/sanjay/repo/pymodbus/examples/server_sync.py", line 295, in <module>
    run_server()
  File "/Users/sanjay/repo/pymodbus/examples/server_sync.py", line 176, in run_server
    StartSerialServer(
  File "/Users/sanjay/repo/pymodbus/pymodbus/server/sync.py", line 750, in StartSerialServer
    server = ModbusSerialServer(context, framer, identity=identity, **kwargs)
  File "/Users/sanjay/repo/pymodbus/pymodbus/server/sync.py", line 583, in __init__
    self._build_handler(kwargs.get("handler", CustomSingleRequestHandler))
  File "/Users/sanjay/repo/pymodbus/pymodbus/server/sync.py", line 613, in _build_handler
    self.handler = handler(request, (self.device, self.device), self)
TypeError: 'NoneType' object is not callable
```

File: examples/client_sync.py
Comment (janiversen) [reply examples/client_sync.py:125]: That might be a good reason to look in ModbusTcpClient and not adapt the example to something false.

it is solved on dev_client (which have other problems).

File: examples/server_sync.py
Comment (dhoomakethu) [reply examples/server_sync.py:288]: This is a slight typo, I will have it fixed. I am missing the inner `list` inside. The problem with `port` expected to  be int is that I can not pass a port of choice when trying to spin a serial server . If that is what is expected then there is no point of having a command line option to set port.

File: examples/server_sync.py
Comment (janiversen) [reply examples/server_sync.py:183]: then the correct way would be to solve it in sync.py

File: pymodbus/client/base.py
Comment (dhoomakethu) [reply pymodbus/client/base.py:81]:
```
/Users/sanjay/.virtualenvs/pymodbus/bin/python /Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py --com tcp --framer socket --port 5020
22:35:30 INFO  client_sync:47 ### Create client object
22:35:30 INFO  client_sync_basic_calls:16 ### Reading Coil
Traceback (most recent call last):
  File "/Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py", line 119, in <module>
    run_sync_client(demonstrate_calls)
  File "/Users/sanjay/repo/pymodbus/examples/client_sync.py", line 124, in run_sync_client
    modbus_calls(client)
  File "/Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py", line 112, in demonstrate_calls
    handle_coils(client)
  File "/Users/sanjay/repo/pymodbus/examples/client_sync_basic_calls.py", line 17, in handle_coils
    rr = client.read_coils(1, 1, unit=UNIT)
  File "/Users/sanjay/repo/pymodbus/pymodbus/client/mixin.py", line 103, in read_coils
    return self.execute(request)  # pylint: disable=no-member
  File "/Users/sanjay/repo/pymodbus/pymodbus/client/base.py", line 178, in execute
    return self.transaction.execute(request)
  File "/Users/sanjay/repo/pymodbus/pymodbus/transaction.py", line 165, in execute
    response, last_exception = self._transact(
  File "/Users/sanjay/repo/pymodbus/pymodbus/transaction.py", line 289, in _transact
    size = self._send(packet)
  File "/Users/sanjay/repo/pymodbus/pymodbus/transaction.py", line 341, in _send
    return self.client.framer.sendPacket(packet)
  File "/Users/sanjay/repo/pymodbus/pymodbus/framer/__init__.py", line 43, in sendPacket
    return self.client.send(message)  # pylint: disable=no-member
AttributeError: 'NoneType' object has no attribute 'send'
```

File: examples/server_sync.py
Comment (janiversen) [reply examples/server_sync.py:288]: there are a reason,  server_sync is common for multiple transports and the command line is shared.

**Comment by dhoomakethu** [comment]
I do not know how do you test,  for me I am having a lot of issues and I am trying to solve. May be you are too fast with your changes that by the time I am finding these bugs you have already fixed it.

File: pymodbus/client/base.py
Comment (janiversen) [reply pymodbus/client/base.py:81]: As I said before, you are using dev, which right now is unstable...if you want something stable, use an older version.

Right now you using my time to answer your messages instead of getting the problems solved.

File: examples/server_async.py
Comment (dhoomakethu) [reply examples/server_async.py:289]: I don't see any complication here, use `port` if passed or use the comm defaults. What is so complicated here ?

**Comment by janiversen** [comment]
Or maybe you do not accept that dev from to time is unstable, that is the nature of dev.

Of course you know how I test, I wrote it to you a little while ago. Green light from CI is test, examples will be solved when the client is stabilised.

**Comment by janiversen** [comment]
Again may I suggest you use an older version of dev, then you can work on what you like in peace. I expect dev to be stable again within a couple of days.

File: examples/client_sync.py
Comment (dhoomakethu) [reply examples/client_sync.py:125]: Also, 
```
✗ pylint examples/
************* Module examples.client_sync
examples/client_sync.py:125:4: E1101: Instance of 'ModbusTcpClient' has no 'stop' member (no-member)
examples/client_sync.py:125:4: E1101: Instance of 'ModbusUdpClient' has no 'stop' member (no-member)
examples/client_sync.py:125:4: E1101: Instance of 'ModbusSerialClient' has no 'stop' member (no-member)
examples/client_sync.py:125:4: E1101: Instance of 'ModbusTlsClient' has no 'stop' member (no-member)

-----------------------------------
````

Comment (janiversen) [reply examples/client_sync.py:125]: Yes it is actually the same problem, if you just dig a little.

Comment (janiversen) [reply examples/client_sync.py:125]: funny enough the tests did not encounter this problem.

**Comment by dhoomakethu** [comment]
> I thought you wrote something about "Don't want to step in to your refactoring work." !
> 
> But seems you have decided not to follow that.
> 
> It is nice to see you active again, and a larger number of your changes are good....BUT some of the changes are made for a purpose (like using comm_defaults) and it is less nice to have those changed.

@janiversen  I don't know whats the problem here, I see an issue and I am trying to fix since I know the code base, I still have the PR open for comments and not yet merged it.  I have pasted all the tracebacks which made me to work on this patch for your reference.  I appreciate you taking time to make this project better but at the same time I request you to be open for the changes.

**Comment by janiversen** [comment]
I am open for changes, look at PR I have reviewed and merged from others!

But when you sidestep (like removing .stop) instead of solving the problem, that is not very helpful. And hopefully something you would also note if you saw me do the same.

File: examples/client_sync.py
Comment (dhoomakethu) [reply examples/client_sync.py:125]: There was a `connect` before and I don't remember seeing a `stop` , So I got the example working for now. When the client is fixed, we can add the `stop` back. Again I have no strong opinions on having an explicit `start` and `stop` . The client should handle it automatically.

File: examples/server_sync.py
Comment (dhoomakethu) [reply examples/server_sync.py:183]: As I said in the earlier message, I do not have a clear idea of what are the kind of changes you are planning to bring in.  I am trying to avoid the client/server code base as much as possible and trying to work around the examples.

File: examples/server_sync.py
Comment (dhoomakethu) [reply examples/server_sync.py:288]: Exactly, then it should be intuitive enough to be used across serial and tcp based protocols with out much fuss.

File: examples/client_sync.py
Comment (janiversen) [reply examples/client_sync.py:125]: I do have a strong opinion. You should not do IO operations in __init__ or __new__.

Making a start(), stop() allows async handling.

File: pymodbus/client/base.py
Comment (dhoomakethu) [reply pymodbus/client/base.py:81]: I am having my problems solved on my own.

Comment (dhoomakethu) [reply pymodbus/client/base.py:81]: > I am open for changes, look at PR I have reviewed and merged from others!
> 
> But when you sidestep (like removing .stop) instead of solving the problem, that is not very helpful. And hopefully something you would also note if you saw me do the same.

I am reiterating it again, I do not want to step on your refactoring and thats the reason I am having these changes done at the example level for the time being. We can revert examples easily. Consider this as a testing I am doing.

**Comment by dhoomakethu** [comment]
I am going to close this PR since it doesn't go well with the other maintainer here.  In my opinion CI green light is not enough, we should also make sure the examples actually work.  `dev` to be unstable is expected but having it completely  unusable is unexpected. May I suggest we move the default branch to something `stable` for the time being.

**Comment by janiversen** [comment]
If you automate the examples, we can include them in CI, and thus have them tested every time...that is actually a very good idea.

You will see some of the examples, have not been working for a very long time (actually long before I started making changes).

File: examples/client_sync.py
Comment (dhoomakethu) [reply examples/client_sync.py:125]: Yes, that's your opinion and we can always agree to disagree.  async handling should done at `read` or `write` transaction layer in my opinion.  Something like this.
```
c = ModbusAsyncClient(....)
await c.read_holding_registers(....)
```

Comment (janiversen) [reply examples/client_sync.py:125]: ???? how about connect() and close() which also are IO.

Comment (janiversen) [reply examples/client_sync.py:125]: Do you really mean it is ok to have a __init_ hanging a long time just to connect().

Comment (dhoomakethu) [reply examples/client_sync.py:125]: `start` and `stop` are must for servers and really to be made mandatory for clients. We can start new discussion to understand each other's thoughts better. I believe there is a confusion here.

**Comment by janiversen** [comment]
I have cherry-picked the real problems you found, thanks.

The are currently available on dev_client, which I hope to merge in a couple of days. There are 1 test outstanding and testing the examples (only those directly in examples).

**Pull request closed without merging**
